# Practice a number of grouping (similar to GROUP command in SQL) techniques on your project data in pandas

# 1. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
# Importing libraries
import pandas as pd 
import numpy as np
import os

Importing and then double-checking imported data

In [2]:
# Defining folder path
path = r'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [3]:
# Checking folder path
path

'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis'

In [4]:
# importing "orders_products_merged_updated.pkl" and calling it "ords_prods_merge"
ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [5]:
# Creating a subset called df of the first one million entries
df = ords_prods_merge[:1000000]

In [6]:
# Checking shape and dimensions of newly created df
df.shape

(1000000, 19)

In [7]:
# Double checking the column names and first couple of rows
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_previous_order  product_id  \
0                  8                        NaN         196   
1                  7                       15.0         196   
2                 12                       21.0         196   
3                  7                       29.0         196   
4                 15                       28.0         196   
5                  7                       19.0         196   
6                  9                       20.0         196   
7                 14                       14.0         196   
8                 16                        0.0         196   
9                  8                       30.0         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   
5                  1          1         Soda        77              7     9.0   
6                  1          1         Soda        77              7     9.0   
7                  2          1         Soda        77              7     9.0   
8                  4          1         Soda        77              7     9.0   
9                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy    Slowest days   
2   both  Mid-range product  Regularly busy    Slowest days   
3   both  Mid-range product      Least busy    Slowest days   
4   both  Mid-range product      Least busy    Slowest days   
5   both  Mid-range product  Regularly busy  Regularly busy   
6   both  Mid-range product  Regularly busy    Busiest days   
7   both  Mid-range product  Regularly busy    Busiest days   
8   both  Mid-range product  Regularly busy    Busiest days   
9   both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  
5        Average orders  
6           Most orders  
7           Most orders  
8           Most orders  
9        Average orders

# Grouping Data with pandas

# Grouping data follows a set of logical steps:

# 1. Split the data into groups based on the given criteria

# 2. Apply a function to these groups

# 3. Combine the results to see what you've just created (group a given subset of a df by a given column--in this case, "product_name" column)

In [8]:
# Using newly created subset
# and using group(by)to group a given dataframe by a given column
df.groupby('product_name')

But, what happened to the output if executed successfully?

By itself, the groupby() function is a little worthless in terms of output because the output of the groupby() function isn’t visible. 

Though it executed successfully (as evidenced by the output message you received), it’s not something that you, as the coder, can actually see. 

It needs you to do something else, for instance, aggregate the data or apply a function (the second step of the workflow), before you’ll be able to see the results.

# As a result, and so far, you’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column). 

# Now, let’s take a look into the second step, which will involve some aggregation!

# Aggregating Data with agg() - concept of pivot tables in Excel

Aggregation functions create summaries of selected columns and store these summary values in new columns. These summaries usually take the form of a descriptive statistic, for instance, a mean, median, maximum, or minimum—the very concept that you encountered in pivot tables in Excel. These aggregations though are important not only for descriptive purposes, but also as a standardization for data to be used for predictive modelling (it reduces the amount of random variation in the data). 

# Instacart Project agg data insights

If you were to calculate the mean of the “order_number” column grouped by the “department_id” column, you could quickly compare the average number of orders per user across each Instacart department. 

(Recall that “order_number” refers to the number of orders placed by a given user.) 

Let’s see how you might use this function to produce a single descriptive statistic for the “order_number” column. This will involve two of the steps in the three-step process:

# 1. Split the data into groups based on “department_id.”

# 2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

The table should list all the possible “department_id” values from the current dataframe together with their corresponding “order_number” means—in other words, the average number of orders per user for each department ID. 

You can easily see, for instance, that the department with an ID of 4 (produce) has a mean of around 19. 

Conversely, the department with an ID of 17 (household), only has a mean of around 11. 

# The insight you can glean from this, then, is that produce sells considerably more on average than household goods.

However, do note one thing—you shouldn’t take everything at face value. Back in the first Achievement, you learned that there are designated statistical tests for proving a significant difference between the means of different groups (e.g., the t-test) because they often can’t be spotted with the naked eye.


There are some aggregations that can be conducted without use of the agg() function. For instance, the command above could be replaced with a command that uses the mean() function to achieve the same results. It could be good to apply such a test here to confirm the statistical significance of any differences in mean values.

Just remember the key difference in syntax between the two methods: when using agg(), put the column you want to aggregate inside the parentheses of the agg() function as an argument. 

When using mean() (or any other standard aggregation function), simply index the column with square brackets, then follow it with the function you want to use after the dot.

In [9]:
# Performing a single aggregation 
# groupby() function is being assigned to the df dataframe creating the pandas object for "department_id"
# agg() function returns the "descriptive stat--mean" of a given column
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [10]:
# Performing a single aggregation for a single stat
# the different format mean() function is used for same results
# conducted without use of agg() function
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

NOTE: Dot notatation vs. square brackets not recommended as not known whether you are indexing or applying a function

-and square brackets are for indexing only

-and if using dot for indexing as well, can make python slower

In [11]:
# Using dot notation vs. square brackets
# for same output of results as the function with brackets
df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

# Performing Multiple Aggregations

In [12]:
# Performing a multiple agg including additional arguments
# producing multiple stats (in new additional given columns)
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Instacart Project additional multiple aggs insights cont...

Each of stats is represented by an additional column in the output, however, the results do not provide additional insights because the min orders will always be 1 and the max 99, capped by data engineers at IC.

# Aggregating Data with transform() - the last step of the process by using the results to create a flag

# Now, let’s map this task onto the three-step process introduced earlier:

1. Split the data into groups based on the “user_id” column.


2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.


3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

Once this process is complete, you can use the “max_order” column to create another new column that assigns a loyalty flag to each customer according to the criteria (via the loc() function).

# Creating a loyalty flag column in "ords_prods_merge" df

First, a new column called “max_order” is created, which will be what stores the maximum order number for each user (step 3). 

Then, the ords_prods_merge dataframe is grouped by the “user_id” column (step 1). 

And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2).

But what is this np? This is actually the NumPy library! The max() function is a function included within NumPy that finds the max value within a column. Including this as an argument within the transform() function tells Python to “transform the ‘order_number’ column by applying the max() function from the NumPy library.”

In [13]:
# Creating a loyalty flag column using loc in "ords_prods_merge" df
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [14]:
#Checking to see new max_order column name and rows
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   

    order_hour_of_day  days_since_previous_order  product_id  \
0                   8                        NaN         196   
1                   7                       15.0         196   
2                  12                       21.0         196   
3                   7                       29.0         196   
4                  15                       28.0         196   
5                   7                       19.0         196   
6                   9                       20.0         196   
7                  14                       14.0         196   
8                  16                        0.0         196   
9                   8                       30.0         196   
10                  9                        7.0         196   
11                 16                        8.0         196   
12                 11                        7.0         196   
13                  9                        7.0         196   
14                 10                       14.0         196   

    add_to_cart_order  reordered product_name  aisle_id  department_id  \
0                   1          0         Soda        77              7   
1                   1          1         Soda        77              7   
2                   1          1         Soda        77              7   
3                   1          1         Soda        77              7   
4                   1          1         Soda        77              7   
5                   1          1         Soda        77              7   
6                   1          1         Soda        77              7   
7                   2          1         Soda        77              7   
8                   4          1         Soda        77              7   
9                   1          1         Soda        77              7   
10                  2          0         Soda        77              7   
11                  6          1         Soda        77              7   
12                  1          1         Soda        77              7   
13                  1          1         Soda        77              7   
14                  1          1         Soda        77              7   

    prices _merge    price_range_loc     busiest_day    busiest_days  \
0      9.0   both  Mid-range product  Regularly busy  Regularly busy   
1      9.0   both  Mid-range product  Regularly busy    Slowest days   
2      9.0   both  Mid-range product  Regularly busy    Slowest days   
3      9.0   both  Mid-range product      Least busy    Slowest days   
4      9.0   both  Mid-range product      Least busy    Slowest days   
5      9.0   both  Mid-range product  Regularly busy  Regularly busy   
6      9.0   both  Mid-range product  Regularly busy    Busiest days   
7      9.0   both  Mid-range product  Regularly busy    Busiest days   
8      9.0   both  Mid-range product  Regularly busy    B

# Instacart Project additional  agg data with transform() insights cont...

The role of flags helps you target specific groups within your data according to a set of conditions, allowing for segmented analysis.

“Loyalty” customers are those that are loyal to a brand. These are customers that come back time and time again and use the service or buy the product often. 

You’ve been asked to locate these loyalty customers at Instacart so that the business strategy team can employ some kind of bonus point program for them. 

You’ll do this by first using the transform() function, which will create a new column containing the maximum frequency of the “order_number” column, and then using the loc() function, which will create a second column containing a flag designating whether a customer is “loyal” or not.

# A new “max_order” column appeared at the end of your dataframe. Each value in this column corresponds to the maximum number of orders made by each user ID.

Of KEY NOTE here is the fact that each discrete value within the “max_order” column will appear in the dataframe the same number of times as its value. For instance, if the max order is 10, then that 10 will appear within the dataframe 10 times—once for each time that user ID made an order.

By comparing the number of orders with the order number and the max order value, you can ascertain the accuracy of your aggregation. You can check these values by scrolling through the “user_ID” column and matching the max order value to the number of times the user ID appears within the dataframe. For the user with an ID of 1, for instance, you can see that the numbers in the “order_number” column increase up to 10. This matches the value in the “max_order” column, so you know your data is accurate.

Another way to conduct this check is to print the head of the dataframe with an argument of, say 100—ords_prods_merge.head(100)

In [15]:
# Checking the accuracy of aggregation
# Printing the output of the head of the df
# with an argument of a certain value (100)
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_previous_order  product_id  \
0                   8                        NaN         196   
1                   7                       15.0         196   
2                  12                       21.0         196   
3                   7                       29.0         196   
4                  15                       28.0         196   
..                ...                        ...         ...   
95                 12                        NaN         196   
96                 17                        3.0         196   
97                 10                       22.0         196   
98                 19                       30.0         196   
99                 19                       16.0         196   

    add_to_cart_order  reordered product_name  aisle_id  department_id  \
0                   1          0         Soda        77              7   
1                   1          1         Soda        77              7   
2                   1          1         Soda        77              7   
3                   1          1         Soda        77              7   
4                   1          1         Soda        77              7   
..                ...        ...          ...       ...            ...   
95                  1          0         Soda        77              7   
96                  9          0         Soda        77              7   
97                  3          0         Soda        77              7   
98                  2          0         Soda        77              7   
99                  2          1         Soda        77              7   

    prices _merge    price_range_loc     busiest_day    busiest_days  \
0      9.0   both  Mid-range product  Regularly busy  Regularly busy   
1      9.0   both  Mid-range product  Regularly busy    Slowest days   
2      9.0   both  Mid-range product  Regularly busy    Slowest days   
3      9.0   both  Mid-range product      Least busy    Slowest days   
4      9.0   both  Mid-range product      Least busy    Slowest days   
..     ...    ...                ...             ...             ...   
95     9.0   both  Mid-range product  Regularly busy  Regularly busy   
96     9.0   both  Mid-range product  Regularly busy  Regularly busy   
97     9.0   both  Mid-range product      Least busy    Slowest days   
98     9.0   both  Mid-range product  Regularly busy    Busiest days   
99     9.0   both  Mid-range product  Regularly busy    Slowest days   

   busiest_period_of_day  max_order  
0         Average orders         10  
1         Average orders         10  
2            Most orders         10  
3         Average orders         10  
4            Most orders         10  
..                   ...        ...  
95           Most orders          3  
96        Average orders          3  
97           Most orders          8  
98        Average orders         22  
99        Average orders         22  

[100 rows x 20 columns]

# What's with showing only the empty and limited rows...(and not the 100 you asked for)?

# You can simply delete that output later by navigating to the Cell menu → Current Outputs → Clear.

This is because Jupyter’s display option is currently set to prevent you from printing too many rows via the head() function. Right now, it’s getting in the way of checking your aggregation procedure. Fortunately, you can change this:

pd.options.display.max_rows = None

This command tells pandas not to assign any options regarding the maximum number of rows to display. After executing this line of code, try your head() function again. You should now see all 100 rows, which will allow you to check whether your aggregation procedure was conducted successfully. Keep in mind that having an output this large in your Jupyter notebook will inflate its size significantly!

# Now that you’ve created your new column, it’s time to set your flag. And for that, it’s back to the loc() function--

# Deriving Columns with loc()

# With your new column ready to go, all that’s left is to create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value.

To create your flag, you’ll need some criteria. You can use the following: however, in most cases, these criteria will be provided by your client, so you won’t need to come up with them yourself.

If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”

If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”

If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [16]:
# creating and setting a first flag that assigns a “loyalty” label to a user ID 
# based on its corresponding max order value--"Loyal customer"
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
# creating and setting a second flag that assigns a “loyalty” label to a user ID 
# based on its corresponding max order value--"Regular customer"
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [18]:
# creating and setting a third flag that assigns a “loyalty” label to a user ID 
# based on its corresponding max order value--"New customer"
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
# Printing the frequency of your new "loyalty_flag" column using vale_counts()
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: count, dtype: int64

Observation: Looks like the majority of customers fall into the “Regular Customer” category, which makes sense! 

# Now, ALWAYS CHECK that everything was interpreted correctly and that the right flags were assigned by accessing multiple columns at the same time. 

Plus, you can use indexing and head() function without looking at the entire df that is prone to and could result in error.

The first set of brackets is the same as the brackets used in df['column'] above

The second set of brackets is indicating that what you’re indexing is a list of multiple columns (and lists are always included within brackets). This is why there are two sets of brackets.


The code itself will display the first 25 rows of just the “user_id,” “loyalty_flag,” and “order_number” columns, which will make it much easier for you to quickly confirm the numbers and ensure accuracy of the flag labels.

In [21]:
# Checking the accuracy of the "loyalty_flag" column
# With accessing output from said multiple columns at the same time
# In this case; "user_id", "loyalty_flag", and "order_number" columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(25)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2

# And that brings you to the end of the three-step process---the basic workflow as a DA! 

# You started by grouping your data-- and learned how grouping and aggregating your data can supplement variable derivation 

# followed up by applying an aggregation function--which can be used to return a variety of descriptive statistics for use in your analysis

# and finished everything off by using the resulting aggregation to create a flag column.

# 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.


In [22]:
# Performing a single aggregation 
# groupby() function is being assigned to the entire df creating the pandas object for "department_id"
# agg() function returns the "descriptive stat--mean" of a given "order_number" column
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

By calculating the mean of the:

"order_number = number of orders placed by a given user " column, grouped by the "department_id" column;  

You could then compare the average number of orders per user across each IC department:

Example:

SUBSET:1M df: department_id = 4 = produce with mean = 18.83
ENTIRE df:    department_id = 4 = produce with mean = 17.81

Conversely :    
SUBSET:1M df: department_id = 17 = household with mean = 11.29
ENTIRE df: department_id = 17 = household with mean = 15.69


As results indicate above, there is less standard deviation from the mean when calculating based on an entire (complete data) df rather than a subset (smaller data) df. The "order_number mean" results from the subset:1M are greater than the "order_number mean" results from the entire df, with the exception of department_id = 17. 

Then, there are designated statistical tests for proving a significant difference between the means of different groups (e.g., the t-test) because they often can’t be spotted with the naked eye. 
 
Further, when using a subset of day, the entire data as a whole is not necessarily a full representation. However, you can deduce, then, other insights such as is produce sells considerably more on average than household goods for example.

The department_id = 21 has a MAX mean of 22.90 which is representative of the largest number of orders placed by a given user, followed by department_id = 10 = 20.20 in second. 
If you are only calculating a SUBSET:1M, department_id = 21 and = 10 do not appear in the output and therefore are shown as "non-existent" and/or "missing" values, and you could not provide further useful comparisons.


SUBSET:1M df: "missing" department_id = 21, therefore we cannot "see" a MAX mean, rather it appears department_id = 16 has the MAX mean = 19.46
ENTIRE df:    department_id = 21 = with MAX mean = 22.90               

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

# Completed--output [13-21]

# 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [23]:
# Performing a multiple agg including additional arguments
# producing multiple stats (in new additional given columns)
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.388747  1.0  99999.0
New customer      13.294370  1.0  99999.0
Regular customer  12.496203  1.0  99999.0

# 5 A
The "mean" prices of products purchased differ from highest to lowest for the three types of customers as follows: 

New customer = highest = 13.29

Regular customer = higher = 12.50 and

Loyal customer = lowest = 10.39

Each of stats is represented by an additional column in the output, however, the results do not provide additional insights because the min orders will always be 1 and the max 99, capped by data engineers at IC.

# 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:

If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [24]:
# Creating a new "avg_price" column to place results of agg
# grouping the ords_prods_merge dataframe by the “user_id” column 
# applying transform() function on the “prices” column
# with the np.mean argument to generate the average price across all their orders for each user
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [25]:
#Checking to see new avg_price column name and rows
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   

    order_hour_of_day  days_since_previous_order  product_id  \
0                   8                        NaN         196   
1                   7                       15.0         196   
2                  12                       21.0         196   
3                   7                       29.0         196   
4                  15                       28.0         196   
5                   7                       19.0         196   
6                   9                       20.0         196   
7                  14                       14.0         196   
8                  16                        0.0         196   
9                   8                       30.0         196   
10                  9                        7.0         196   
11                 16                        8.0         196   
12                 11                        7.0         196   
13                  9                        7.0         196   
14                 10                       14.0         196   

    add_to_cart_order  reordered  ... department_id  prices  _merge  \
0                   1          0  ...             7     9.0    both   
1                   1          1  ...             7     9.0    both   
2                   1          1  ...             7     9.0    both   
3                   1          1  ...             7     9.0    both   
4                   1          1  ...             7     9.0    both   
5                   1          1  ...             7     9.0    both   
6                   1          1  ...             7     9.0    both   
7                   2          1  ...             7     9.0    both   
8                   4          1  ...             7     9.0    both   
9                   1          1  ...             7     9.0    both   
10                  2          0  ...             7     9.0    both   
11                  6          1  ...             7     9.0    both   
12                  1          1  ...             7     9.0    both   
13                  1          1  ...             7     9.0    both   
14                  1          1  ...             7     9.0    both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy        Average orders   
1   Mid-range product  Regularly busy    Slowest days        Average orders   
2   Mid-range product  Regularly busy    Slowest days           Most orders   
3   Mid-range product      Least busy    Slowest days        Average orders   
4   Mid-range product      Least busy    Slowest days           Most orders   
5   Mid-range product  Regularly busy  Regularly busy        Average orders   
6   Mid-range product  Regularly busy    Busiest days           Most orders   
7   Mid-range product  Regularly busy    Busiest days           Most orders   
8   Mid-range product  Regularly busy    B

In [26]:
# creating and setting a first flag that assigns a “spending” label to a user ID 
# based on its corresponding "lower than 10 value"--"Low spender"
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [27]:
# creating and setting a second flag that assigns a “spending” label to a user ID 
# based on its corresponding "greater than or equal to 10 value"--"High spender"
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [31]:
#Checking to see new spending_flag column name and rows
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_previous_order  product_id  \
0                   8                        NaN         196   
1                   7                       15.0         196   
2                  12                       21.0         196   
3                   7                       29.0         196   
4                  15                       28.0         196   
..                ...                        ...         ...   
95                 12                        NaN         196   
96                 17                        3.0         196   
97                 10                       22.0         196   
98                 19                       30.0         196   
99                 19                       16.0         196   

    add_to_cart_order  reordered  ... prices  _merge    price_range_loc  \
0                   1          0  ...    9.0    both  Mid-range product   
1                   1          1  ...    9.0    both  Mid-range product   
2                   1          1  ...    9.0    both  Mid-range product   
3                   1          1  ...    9.0    both  Mid-range product   
4                   1          1  ...    9.0    both  Mid-range product   
..                ...        ...  ...    ...     ...                ...   
95                  1          0  ...    9.0    both  Mid-range product   
96                  9          0  ...    9.0    both  Mid-range product   
97                  3          0  ...    9.0    both  Mid-range product   
98                  2          0  ...    9.0    both  Mid-range product   
99                  2          1  ...    9.0    both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy        Average orders        10   
1   Regularly busy    Slowest days        Average orders        10   
2   Regularly busy    Slowest days           Most orders        10   
3       Least busy    Slowest days        Average orders        10   
4       Least busy    Slowest days           Most orders        10   
..             ...             ...                   ...       ...   
95  Regularly busy  Regularly busy           Most orders         3   
96  Regularly busy  Regularly busy        Average orders         3   
97      Least busy    Slowest days           Most orders         8   
98  Regularly busy    Busiest days        Average orders        22   
99  Regularly busy    Slowest days        Average orders        22   

        loyalty_flag  avg_price  spending_flag  
0       New customer   6.367797    Low spender  
1       New customer   6.367797    Low spender  
2       New customer   6.367797    Low spender  
3       New customer   6.367797    Low spender  
4       New customer   6.367797    Low spender  
..               ...        ...            ...  
95      New customer  10.006250   High spender  
96      New customer   8.496552    Low spender  
97      New customer   7.396610    Low spender  
98  Regular customer   7.657813    Low spender  
99  Regular customer   7.657813    Low spender  

[100 rows x 23 columns]

In [32]:
# Printing the frequency of your new "spending_flag" column using value_counts()
ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31798751
High spender      635461
Name: count, dtype: int64

# 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [33]:
# Creating a new "median_days_since_previous_order" column to place results of agg
# grouping the ords_prods_merge dataframe by the “user_id” column 
# applying transform() function 
# with the np.mean argument to generate the median for each user
ords_prods_merge['median_previous_order'] = ords_prods_merge.groupby(['user_id'])['days_since_previous_order'].transform(np.median)

In [34]:
# creating and setting a first flag that assigns a “customer order frequency” label to a user ID 
# based on its corresponding "higher than 10 and <=20 value"--"Regular customer"
ords_prods_merge.loc[ords_prods_merge['median_previous_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [36]:
# creating and setting a second flag that assigns a “customer order frequency” label to a user ID 
# based on its corresponding "higher than 20 value"--"Non-frequent customer"
ords_prods_merge.loc[ords_prods_merge['median_previous_order'] > 10 & (ords_prods_merge['median_previous_order']<=20),'order_frequency_flag'] = 'Regular customer'

In [37]:
# creating and setting a third flag that assigns a “customer order frequency” label to a user ID 
# based on its corresponding "lower than or =10 value"--"Frequent customer"
ords_prods_merge.loc[ords_prods_merge['median_previous_order'] <= 10,'order_frequency_flag'] = 'Frequent customer'

In [38]:
#Checking to see new order_frequency_flag column name and rows
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
..       ...      ...      ...           ...                 ...   
95   3226575      360    prior             1                   5   
96   1469869      377    prior             3                   5   
97   1927023      387    prior             2                   4   
98    858092      420    prior             4                   1   
99    119314      420    prior             5                   3   

    order_hour_of_day  days_since_previous_order  product_id  \
0                   8                        NaN         196   
1                   7                       15.0         196   
2                  12                       21.0         196   
3                   7                       29.0         196   
4                  15                       28.0         196   
..                ...                        ...         ...   
95                 12                        NaN         196   
96                 17                        3.0         196   
97                 10                       22.0         196   
98                 19                       30.0         196   
99                 19                       16.0         196   

    add_to_cart_order  reordered  ...    price_range_loc     busiest_day  \
0                   1          0  ...  Mid-range product  Regularly busy   
1                   1          1  ...  Mid-range product  Regularly busy   
2                   1          1  ...  Mid-range product  Regularly busy   
3                   1          1  ...  Mid-range product      Least busy   
4                   1          1  ...  Mid-range product      Least busy   
..                ...        ...  ...                ...             ...   
95                  1          0  ...  Mid-range product  Regularly busy   
96                  9          0  ...  Mid-range product  Regularly busy   
97                  3          0  ...  Mid-range product      Least busy   
98                  2          0  ...  Mid-range product  Regularly busy   
99                  2          1  ...  Mid-range product  Regularly busy   

      busiest_days  busiest_period_of_day max_order      loyalty_flag  \
0   Regularly busy         Average orders        10      New customer   
1     Slowest days         Average orders        10      New customer   
2     Slowest days            Most orders        10      New customer   
3     Slowest days         Average orders        10      New customer   
4     Slowest days            Most orders        10      New customer   
..             ...                    ...       ...               ...   
95  Regularly busy            Most orders         3      New customer   
96  Regularly busy         Average orders         3      New customer   
97    Slowest days            Most orders         8      New customer   
98    Busiest days         Average orders        22  Regular customer   
99    Slowest days         Average orders        22  Regular customer   

    avg_price spending_flag median_previous_order  order_frequency_flag  
0    6.367797   Low spender                  20.5      Regular customer  
1    6.367797   Low spender                  20.5      Regular customer  
2    6.367797   Low spender                  20.5      Regular customer  
3    6.367797   Low spender                  20.5      Regular customer  
4    6.367797   Low spender                  20.5      Regular customer  
..        ...           ...                   ...                   ...  
95  10.006250  High spender                   4.0     Frequent customer  
96   8.496552   Low spender         

In [39]:
# Printing the frequency of your new "order_frequency_flag" column using value_counts()
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer    21577409
Regular customer     10856803
Name: count, dtype: int64

# 8. Ensure your notebook is clean and structured and that your code is well commented.

# 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [40]:
# Defining new folder pathgrouped
pathgrouped = r'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis/02 Data/Prepared Data'

In [41]:
# Checking if it'll work!
pathgrouped

'/Users/renataherrera/Documents/CF RH 2023-2024/CF DATA IMMERSION/CF RH A4 PYTHON/RH_PYTHON_Instacart Basket Analysis/02 Data/Prepared Data'

In [42]:
# Yup! Go pickle!
ords_prods_merge.to_pickle(os.path.join(pathgrouped,'orders_products_aggregated.pkl'))